server/matchmaker/algorithm/team_matchmaker.py

```diff
@@ -0,0 +1,262 @@
+    7. remove all games with match quality below threshold then sort by quality descending
+    8. pick the first game from the game list and remove all other games that contain the same players
+    9. repeat until the list is empty
+
+    for the future:
+    Find combination of potential games that allows for maximal number of games to launch
+    Repeat 8. with an increasing amount of top games removed beforehand to get all possible game combinations
+    If there is more than one solution, pick the one with highest total game quality
+    Optimization: sort by game quality first, abort when you find a solution with
+    the full number of theoretically possible games (floor(playersInQueue/(teamSize*2)))
+    """
+
+    def find(self, searches: Iterable[Search]) -> List[Match]:
+        searches = list(searches)
+        searches.sort(key=lambda s: s.average_rating, reverse=True)
+        self._logger.debug("=== starting matching algorithm ===")
+        possible_games = set()
+        for index, search in enumerate(searches):
+            self._logger.debug("building game for %s", repr(search))
+            participants = self._pick_neighboring_players(searches, index)
+            try:
+                match = self.make_teams(list(participants))
+                game = self.calculate_game_quality(match)
+                possible_games.add(game)
+            except AssertionError:
+                self._logger.warning("failed to assign even teams. Skipping this game...")
+        self._logger.debug("got %i games", len(possible_games))
+        for game in possible_games:
+            self._logger.debug("game: %s vs %s rating disparity: %i quality: %f",
+                               repr(game.match[0]),
+                               repr(game.match[1]),
+                               game.match[0].cumulated_rating - game.match[1].cumulated_rating,
+                               game.quality)
+        return self._pick_best_noncolliding_games(list(possible_games))
+
+    def _pick_neighboring_players(self, searches: List[Search], index: int) -> List[Search]:
+        participants = []
+        i = 0
+        number_of_players = 0
+        out_of_bounds_counter = 0
+        while number_of_players < self.team_size * 2 and out_of_bounds_counter < 2:
+            try:
+                candidate = searches[index]
+                out_of_bounds_counter = 0
+                if number_of_players + len(candidate.players) <= self.team_size * 2:
+                    participants.append(candidate)
+                    number_of_players += len(candidate.players)
+            except IndexError:
+                out_of_bounds_counter += 1
+            i += 1
+            index += i * pow(-1, i)
+        return participants
+
+    def make_teams(self, searches: List[Search]) -> Tuple[Search, Search]:
+        avg = CombinedSearch(*searches).average_rating
+        team_target_strength = CombinedSearch(*searches).cumulated_rating / 2
```

Comment:
Not a fan of using the search objects in this way, just compute the values directly with generator expressions:
```suggestion
        avg = statistics.mean(search.avg_rating for search in searches)
        team_target_strength = sum(search.cumulated_rating for search in searches) / 2
```

Reply:
This won't work for the average rating, because we need to take the different number of players in the searches into account.

Reply:
Well, write a function that does that then. But creating a new search object just to access one property which will turn into a function call through python magic is not good. You could even have the property call your function instead so the code isn't duplicated.